pkg/server/plugin/upstreamca/awscertificatemanager/awscertificatemanager.go

```diff
@@ -0,0 +1,226 @@
+func (m *CertificateManagerPlugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (*spi.ConfigureResponse, error) {
+	config, err := m.validateConfig(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the client
+	m.certificateManagerClient, err = m.hooks.newClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Perform a check for the presence of the CA in ACM
+	describeResponse, err := m.certificateManagerClient.DescribeCertificateAuthority(&acmpca.DescribeCertificateAuthorityInput{
+		CertificateAuthorityArn: aws.String(config.CertificateAuthorityARN),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure the CA is set to ACTIVE
+	caStatus := aws.StringValue(describeResponse.CertificateAuthority.Status)
+	if caStatus != "ACTIVE" {
+		return nil, fmt.Errorf("certificate authority is in invalid state: %s", caStatus)
+	}
+
+	// Add remaining values to plugin
+	m.certificateAuthorityArn = config.CertificateAuthorityARN
+	m.signingAlgorithm = config.SigningAlgorithm
+	m.caSigningTemplateArn = config.CASigningTemplateARN
+
+	return &spi.ConfigureResponse{}, nil
+}
+
+// GetPluginInfo returns information about this plugin to Spire server
+func (*CertificateManagerPlugin) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.GetPluginInfoResponse, error) {
+	return &spi.GetPluginInfoResponse{}, nil
+}
+
+// SubmitCSR submits the CSR to ACM PCA to be signed by the certificate authority
+func (m *CertificateManagerPlugin) SubmitCSR(ctx context.Context, request *upstreamca.SubmitCSRRequest) (*upstreamca.SubmitCSRResponse, error) {
+	csrBuf := new(bytes.Buffer)
+	err := pem.Encode(csrBuf, &pem.Block{
+		Type:  csrRequestType,
+		Bytes: request.Csr,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info("Submitting CSR with signing algorithm: ", m.signingAlgorithm)
+
+	validityPeriod := time.Second * time.Duration(request.PreferredTtl)
+	issueResponse, err := m.certificateManagerClient.IssueCertificate(&acmpca.IssueCertificateInput{
+		CertificateAuthorityArn: aws.String(m.certificateAuthorityArn),
+		SigningAlgorithm:        aws.String(m.signingAlgorithm),
+		Csr:                     csrBuf.Bytes(),
+		TemplateArn:             aws.String(m.caSigningTemplateArn),
+		Validity: &acmpca.Validity{
+			Type:  aws.String(acmpca.ValidityPeriodTypeAbsolute),
+			Value: aws.Int64(m.hooks.clock.Now().Add(validityPeriod).Unix()),
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Using the output of the `IssueCertificate` call, poll ACM until
+	// the certificate has been issued
+	certificateArn := issueResponse.CertificateArn
+
+	log.WithField("certificateARN", aws.StringValue(certificateArn)).Info("Waiting for issuance from ACM.")
+	getCertificateInput := &acmpca.GetCertificateInput{
+		CertificateAuthorityArn: aws.String(m.certificateAuthorityArn),
+		CertificateArn:          certificateArn,
+	}
+	err = m.certificateManagerClient.WaitUntilCertificateIssued(getCertificateInput)
```

Comment:
How long can this operation take? Does this require operator intervention? If this is going to take a while, it might be a good idea to wire up `ctx` so the wait can get interrupted if the caller has gone away.

Reply:
Seconds, typically < 5. Of course it depends on the load on ACM in a given region at a given point in time, so I'll wire up `ctx`.

Reply:
Added